signac/core/buffered_synced_collection.py

```diff
@@ -0,0 +1,198 @@
+"""Implement the BufferedCollection class.
+
+The BufferedCollection extends the abstract SyncedCollection to support
+transparent buffering of the synchronization process. This feature allows
+client code to indicate to the collection when it is safe to buffer reads and
+writes, which usually means guaranteeing that the synchronization destination
+(e.g. an underlying file or database entry) will not be modified by other
+processes concurrently with the set of operations within the buffered block.
+Judicious use of buffering can dramatically speed up code paths that might
+otherwise involve, for instance, heavy I/O. The specific buffering mechanism
+must be implemented by each backend since it depends on the nature of the
+underlying data format.
+"""
+import logging
+from typing import List, Any
+
+from contextlib import contextmanager
+
+from .errors import BufferedError
+from .synced_collection import SyncedCollection
+
+logger = logging.getLogger(__name__)
+
+_BUFFERED_MODE = 0
+_BUFFERED_BACKENDS: List[Any] = list()
+
+
+@contextmanager
+def buffer_reads_writes():
+    """Enter a global buffer mode for all BufferedCollection instances.
+
+    All future write operations are written to the buffer, read
+    operations are performed from the buffer whenever possible.
+
+    All write operations are deferred until the flush_all() function
+    is called, the buffer overflows, or upon exiting the buffer mode.
+
+    Raises
+    ------
+    BufferException
+    """
+    global _BUFFERED_MODE
+    assert _BUFFERED_MODE >= 0
+
+    _BUFFERED_MODE += 1
+    try:
+        yield
+    finally:
+        _BUFFERED_MODE -= 1
+        if _BUFFERED_MODE == 0:
+            _flush_all_backends()
+
+
+def _flush_all_backends():
+    """Execute all deferred write operations.
+
+    Raises
+    ------
+    BufferedFileError
+    """
+    global _BUFFERED_BACKENDS
+    logger.debug("Flushing buffer...")
+    issues = {}
+    for backend in _BUFFERED_BACKENDS:
+        try:
+            # try to sync the data to backend
+            issue = backend._flush_buffer()
+            issues.update(issue)
+        except OSError as error:
+            logger.error(str(error))
+            issues[backend] = error
+    if issues:
+        raise BufferedError(issues)
+
+
+class BufferedCollection(SyncedCollection):
+    """Adds buffering capabilities to the SyncedCollection.
+
+    *The default behavior is no buffering.* The class simply defines an
+    appropriate interface for buffering behavior so that client code can rely
+    on these methods existing, e.g. to be able to do things like
+    `with collection.buffered...`. However, in the default case the result of
+    this will be a no-op.
+
+    In addition to the methods shown here, we may also require methods to flush
+    a specific entry out of the buffer (e.g. _flush(self, item)) which has been
+    implemented in the current version of the PR. I've left out such details
```

Comment:
This section seems like a TODO that should be marked or converted to `# comments`.